### PR TITLE
Enable native ModelOpt quantization support (2/3)

### PR DESCRIPTION
**Pull request by Edwardf0t1**
This is the second PR in a three-part series to enable native ModelOpt quantization in SGLang. It includes changes from the first PR (https://github.com/sgl-project/sglang/pull/7149) and will be rebased once the first PR is merged.

## Motivation

We aim to enhance SGLang's quantization capabilities, making ModelOpt integration more robust and user-friendly while providing checkpoint persistence for better performance in production environments.

## Modifications

- Created `_setup_modelopt_quantization()` and added calibration functionalities.
- Added `modelopt_checkpoint_restore_path` and `modelopt_checkpoint_save_path` parameters to both `ModelConfig` and `ServerArgs`. These allow users to save and restore quantized checkpoints, avoiding re-quantization on subsequent runs
- Improved error handling during the ModelOpt quantization process.
- Added mode unit tests in `test_modelopt_loader.py` to verify the ModelOpt functionality.

The 3rd PR are also ready for review: https://github.com/sgl-project/sglang/pull/10154 

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by zhyncs** [comment]
hi @Edwardf0t1 can you help fix the conflicts? thanks

**Comment by Edwardf0t1** [comment]
> hi @Edwardf0t1 can you help fix the conflicts? thanks

@zhyncs Just rebased and resolved the conflicts. Could you or @Qiaolin-Yu help review the PR? Thanks.

File: python/sglang/srt/configs/model_config.py [hunk 1]
```diff
@@ -443,36 +452,52 @@ def _parse_quant_hf_config(self):
...
                 )
                 with open(quant_config_file) as f:
                     quant_config_dict = json.load(f)
-                json_quant_configs = quant_config_dict["quantization"]
-                quant_algo = json_quant_configs.get("quant_algo", None)
-                if quant_algo == "MIXED_PRECISION":
-                    quant_cfg = {"quant_method": "w4afp8"}
-                else:
-                    quant_cfg = modelopt_quant_config
+                quant_cfg = self._parse_modelopt_quant_config(quant_config_dict)
```
Comment (jingyu-ml) [inline python/sglang/srt/configs/model_config.py:516]: Is W4AFP8 a quantization format from MO?, if not, i think we can just rename this function to `_parse_quant_config `

File: python/sglang/srt/entrypoints/engine.py [hunk 2]
```diff
@@ -123,6 +123,8 @@ def __init__(self, **kwargs):
         self.port_args = PortArgs.init_new(server_args)
         logger.info(f"{server_args=}")
 
+        print(f"Launching subprocesses with {server_args=}")
```
Comment (jingyu-ml) [inline python/sglang/srt/entrypoints/engine.py]: I don’t think using print is a good approach in distributed inference code.

**Comment by jingyu-ml** [comment]
I think we should add example code in this PR to demonstrate how to use `modelopt_checkpoint_restore_path` or other new functions, so users can understand without needing deep context.

File: python/sglang/srt/model_loader/loader.py [hunk 3]
```diff
@@ -1664,9 +1729,181 @@ def load_model_with_cpu_quantization(
...
+                tokenizer.padding_side = "left"
+
+            from modelopt.torch.utils.dataset_utils import (
+                create_forward_loop,
+                get_dataset_dataloader,
+            )
+
+            # Create calibration dataloader
+            calib_dataloader = get_dataset_dataloader(
+                dataset_name="cnn_dailymail",
```
Comment (jingyu-ml) [inline python/sglang/srt/model_loader/loader.py]: It’s better to make this configurable as TODOs below.

File: python/sglang/srt/model_loader/loader.py [hunk 4]
```diff
@@ -1664,9 +1729,181 @@ def load_model_with_cpu_quantization(
...
+                num_samples=512,  # TODO: Consider making this configurable
+                device=model.device,
+                include_labels=False,
+            )
+
+            calibrate_loop = create_forward_loop(dataloader=calib_dataloader)
+
+            # Apply quantization
+            mtq.quantize(model, quant_cfg, forward_loop=calibrate_loop)
+            mtq.print_quant_summary(model)
```
Comment (jingyu-ml) [inline python/sglang/srt/model_loader/loader.py]: Same as above, only run print_quant_summary on a single rank (print once in rank 0).

File: python/sglang/srt/model_loader/loader.py [hunk 5]
```diff
@@ -474,12 +478,73 @@ def download_model(self, model_config: ModelConfig) -> None:
...
     def load_model(
         self,
         *,
         model_config: ModelConfig,
         device_config: DeviceConfig,
     ) -> nn.Module:
+
+        if hasattr(model_config, "modelopt_quant") and model_config.modelopt_quant:
+            # Load base model using shared method
+            model = self._load_modelopt_base_model(model_config)
```
Comment (jingyu-ml) [inline python/sglang/srt/model_loader/loader.py:568]: I’m a bit confused about this function, what’s the detailed purpose of designing this function? Is the goal to merge all MO-related model loading functions into one, instead of modifying each model’s loading definition separately?

File: python/sglang/srt/configs/model_config.py (hunk 1 above)
Comment (Edwardf0t1) [reply python/sglang/srt/configs/model_config.py:516]: Actually we have w4afp8 recipe in modelopt, e.g., `w4a8_nvfp4_fp8` and `w4a8_mxfp4_fp8`.

File: python/sglang/srt/model_loader/loader.py (hunk 4 above)
Comment (Edwardf0t1) [reply python/sglang/srt/model_loader/loader.py]: Good point, updated.

File: python/sglang/srt/model_loader/loader.py (hunk 3 above)
Comment (Edwardf0t1) [reply python/sglang/srt/model_loader/loader.py]: done

File: python/sglang/srt/entrypoints/engine.py (hunk 2 above)
Comment (Edwardf0t1) [reply python/sglang/srt/entrypoints/engine.py]: Good point, updated with `rank0_log()`

File: python/sglang/srt/model_loader/loader.py (hunk 5 above)
Comment (Edwardf0t1) [reply python/sglang/srt/model_loader/loader.py:568]: The primary purpose is to eliminate code duplication between two different loaders that both need to load models for ModelOpt processing.

The function handles ModelOpt's unique loading requirements such as `accelerate` dependency (for device mapping); empty weights initialization (memory efficiency during quantization); auto device mapping (for distributed/multi-GPU setups); memory management (CPU fallback handling).

There are two usages: 1) `DefaultModelLoader` usage for loading pre-quantized models, and 2) `ModelOptModelLoader` usage for full quantization + export flow.

**Comment by Edwardf0t1** [comment]
> I think we should add example code in this PR to demonstrate how to use `modelopt_checkpoint_restore_path` or other new functions, so users can understand without needing deep context.

The usage is covered in unit tests: `test/srt/model_loader/test_modelopt_loader.py`

File: python/sglang/srt/server_args.py [hunk 6]
```diff
@@ -151,6 +151,13 @@ class ServerArgs:
     load_format: str = "auto"
     model_loader_extra_config: str = "{}"
     trust_remote_code: bool = False
+    dtype: str = "auto"
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/server_args.py]: dtype, kv_cache_dtype, quantization already exist

File: test/srt/model_loader/test_modelopt_loader.py [hunk 7]
```diff
@@ -0,0 +1,810 @@
...
+
+import os
+import sys
+import unittest
+from unittest.mock import MagicMock, patch
+
+import torch.nn as nn
+
+# Add the sglang path for testing
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "../../../python"))
```
Comment (Qiaolin-Yu) [inline test/srt/model_loader/test_modelopt_loader.py]: Check if this is really needed

File: test/srt/model_loader/test_modelopt_loader.py [hunk 8]
```diff
@@ -0,0 +1,810 @@
...
+            )
+
+            # Verify print_quant_summary was called
+            mock_mtq.print_quant_summary.assert_called_once_with(self.mock_base_model)
+
+            # Verify eval() was called on the returned model
+            self.mock_base_model.eval.assert_called()
+
+            # Verify we get back the expected model
+            self.assertEqual(result_model, self.mock_base_model)
```
Comment (Qiaolin-Yu) [inline test/srt/model_loader/test_modelopt_loader.py]: When does `mock_base_model` become `tinyllama`

File: python/sglang/srt/server_args.py (hunk 6 above)
Comment (Edwardf0t1) [reply python/sglang/srt/server_args.py]: Removed.

File: test/srt/model_loader/test_modelopt_loader.py (hunk 7 above)
Comment (Edwardf0t1) [reply test/srt/model_loader/test_modelopt_loader.py]: It's needed to make sure the tests run with the current development branch rather than the default installed sglang library.  
We can also remove it but need to set  `PYTHONPATH=python` when run the tests.

File: test/srt/model_loader/test_modelopt_loader.py (hunk 8 above)
Comment (Edwardf0t1) [reply test/srt/model_loader/test_modelopt_loader.py]: Actually `mock_base_model` never actually becomes `TinyLlama` - it remains a mock object throughout the test. What happens is: 

1. `self.model_path = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"` (just a string path)
2. `self.mock_base_model = MagicMock(spec=nn.Module)` (a mock object that pretends to be a PyTorch module)
3. The TinyLlama model loading is **mocked** with: `with patch.object(loader, "_load_modelopt_base_model", return_value=self.mock_base_model):`

 `self.mock_base_model.eval.return_value = self.mock_base_model` this was configured earlier so `result_model == self.mock_base_model` is `True`.

**Review by jingyu-ml — APPROVED** [review]

File: python/sglang/srt/entrypoints/engine.py [hunk 9]
```diff
@@ -126,6 +126,8 @@ def __init__(self, **kwargs):
         self.port_args = PortArgs.init_new(server_args)
         logger.info(f"{server_args=}")
 
+        logger.info("Launching subprocesses for distributed inference...")
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/entrypoints/engine.py]: Why do we need to add this?

File: python/sglang/srt/model_loader/loader.py [hunk 10]
```diff
@@ -1764,6 +1764,96 @@ def __init__(self, load_config: LoadConfig):
...
+            except Exception as e:
+                logger.warning(
+                    f"Failed to restore from {quantized_ckpt_restore_path}: {e}"
+                )
+                rank0_log("Proceeding with calibration-based quantization...")
+
+        # Set up calibration-based quantization
+        try:
+            # Left padding tends to work better for batched generation with decoder-only LMs
+            with suppress(Exception):
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/model_loader/loader.py:1818]: Why does it need to suppress the exception?

File: python/sglang/srt/model_loader/loader.py [hunk 11]
```diff
@@ -1808,33 +1897,26 @@ def load_model(
...
+                tokenizer,
+                quant_cfg,
+                quantized_ckpt_restore_path=quantized_ckpt_restore_path,
+                quantized_ckpt_save_path=quantized_ckpt_save_path,
+            )
         except Exception as e:
-            logger.error(f"Error during ModelOpt mtq.quantize call: {e}")
-            raise
-        mtq.print_quant_summary(model)
+            logger.warning(f"ModelOpt quantization failed: {e}")
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/model_loader/loader.py:1918]: Why not raise exception here?

File: python/sglang/srt/entrypoints/engine.py (hunk 9 above)
Comment (Edwardf0t1) [reply python/sglang/srt/entrypoints/engine.py]: This was added for debugging. Removed.

File: python/sglang/srt/model_loader/loader.py (hunk 10 above)
Comment (Edwardf0t1) [reply python/sglang/srt/model_loader/loader.py:1818]: Not all tokenizers support the `padding_side` attribute: Some tokenizer implementations may not have this attribute or may not allow it to be modified. If this modification fails, the quantization process should still continue, since it's not essential.

File: python/sglang/srt/model_loader/loader.py (hunk 11 above)
Comment (Edwardf0t1) [reply python/sglang/srt/model_loader/loader.py:1918]: This is in the `ModelOptModelLoader.load_model()` method, which is designed to load a model with optional quantization. If quantization fails, the system falls back to loading the unquantized model rather than completely failing.

**Review by Qiaolin-Yu — APPROVED** [review]
LGTM

**Comment by Edwardf0t1** [comment]
> LGTM

Thanks @Qiaolin-Yu for the review and approval.  
The failed CI tests seem unrelated to the PR. Could we unblock to merge? @zhyncs @Ying1123 @merrymercy
